Code Review for GNU Radio Block Header Parsing tool.

aru31 opened the pull request:
Code Review for GNU Radio Block Header Parsing tool

aru31 committed:
Add base.py file in cli module to import override Click functions

aru31 committed:
Create cli and core base module for AST generation of header blocks

aru31 committed:
Create basic CLI for blocktool with minimal support

aru31 committed:
Add Sequence Completer to CLI and successful generation of AST

aru31 committed:
CLI structure complete with parseheader command

aru31 committed:
Basic core structure complete

aru31 committed:
Add test script gr_blocktool to run the tool

aru31 committed:
Add JSON schema and validation for parsed json output file

aru31 committed:
Change properties and methods key to list in JSON schema

aru31 committed:
Create an independent api from blocktool

aru31 committed:
Bug fix for abslute path of the header files

aru31 committed:
Create basic parser core api

aru31 committed:
Parse the block header documentation

aru31 committed:
Expose the core api, minor bug fixes

aru31 committed:
Create the code pylint compatible

aru31 committed:
Modify cli to accept file_path as an argument, parse default values of make function arguments

aru31 committed:
Fix: Namespace parsing of block header file

aru31 committed:
Parse the io_signature from the implementation file of the block header

aru31 committed:
Create json file generator

aru31 committed:
Add key-value io_signature and docstring in json schema, change sample generated json output

aru31 committed:
Fix: squash an I/O parsing bug

aru31 committed:
Change directory structure for blocktool tests

aru31 committed:
Add Blocktool unittest

aru31 committed:
Removed empty strings, make the code pylint compatible

aru31 committed:
Use str.format() to get output

aru31 committed:
Implement YAML generator

aru31 committed:
Add a new CLI argument to parse a complete header directory

aru31 committed:
Add Logger to log errors without raising exceptions

Commit:
modtool: c++ qa tests cmake error

C++ QA tests were not being added to the correct spot in the
lib/CMakeLists

Fixes #2559

aru31 committed:
Create output schema file in blocktool core

aru31 committed:
Change directory structure of blocktool and cli commands

aru31 committed:
write unittests for Blocktool Excceptions

aru31 committed:
Add sample yaml files

primercuervo commented on gr-utils/python/blocktool/core/iosignature.py:
```diff
@@ -0,0 +1,139 @@
+        for line in impl:
+            if Constants.IO_SIGNATURE in line:
+                lines.append(line)
+    impl.close()
```
When using `with open`, the file gets automatically closed after the `with` block thanks to the context manager, so line 54 is not required.
```suggestion
```

primercuervo commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,278 @@
+    def validate(self):
+        """ Override the Blocktool validate function """
+        BlockTool._validate(self)
+        self._validate()
```
I'm a bit confused about L90-L91. Aren't both lines up to this point supposed to be just `pass`?

aru31 replied:
L90 and 91 points to the same validate function in `core/base.py`, sorry for the repetition. They are a pass, for now, but if blocktool gets additional functionality in future, common blocktool validation in `core/base.py` might come in handy.

primercuervo commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,278 @@
+                            getter_arguments.append(args['name'])
+                            setter_args['arguments_type'].append(args.copy())
+                        self.parsed_data['methods'].append(setter_args.copy())
+        except:
```
Wouldn't it be a bit obscure for the API user if the exception doesn't `raises` or produce a type of input to work with? it seems that when the getters and setters fail we end up, unknowingly, with empty parsed data

aru31 replied:
This was one of the doubts I had and wanted to clear with you. Sure, an exception can be raised.

primercuervo commented on gr-utils/python/blocktool/tests/test_blocktool.py:
```diff
@@ -0,0 +1,187 @@
+        self.gr = self.modname.split('-')[0]
+        self.module = self.modname.split('-')[-1]
+
+    def is_int(self, number):
```
this is a general UnitTest but seems unrelated to the specifics of BlockTool

aru31 replied:
I didn't get it. All the __init__ variables defined above are required to check the Blocktool raised exceptions as in the validate function of `core/parseheader.py`

skoslowski commented on gr-utils/python/blocktool/__init__.py:
```diff
@@ -0,0 +1,23 @@
+#
+
+from .core.parseheader import BlockHeaderParser
+from .core.outputschema import PARSED_OUTPUT_SCHEME
```
I prefer s.th. like RESULT_SCHEMA

skoslowski commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,281 @@
+import codecs
+import logging
+
+import click
```
No cli stuff in the core, please

skoslowski commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,281 @@
+
+from pygccxml import parser
+from pygccxml import declarations
+from pygccxml import utils
```
collapse?

skoslowski commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,281 @@
+        self.info['impl_dir'] = os.path.abspath(os.path.join(file_path,
+                                                             '..', '..', '..', '..', 'lib'))
+        self.info['impl_file'] = os.path.join(self.info['impl_dir'],
+                                              self.info['filename'].split('.')[0]+'_impl.cc')
```
why aren't these defined as attributes of the instance? want benefit does "info" have

skoslowski commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,281 @@
+        try:
+            open(self.info['impl_file'], 'r')
+        except OSError:
+            raise OSError
```
What is this for?

skoslowski commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,281 @@
+                                   'gnuradio',
+                                   self.info['modname'].split('-')[-1])
+        if _target_dir != self.info['target_dir']:
+            raise BlockToolException('Invalid GNU Radio module')
```
Explain why in the message

aru31 replied:
I have now cleaned the exceptions.

skoslowski commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,281 @@
+                            self.parsed_data['namespace'].append(
+                                str(_namespace).split('::')[-1].split(' ')[0])
+        except BlockToolException as exception:
+            raise BlockToolException(
```
This looks weird to me, shouldn't this be another type of exception

aru31 replied:
Changed to RuntimeError

skoslowski commented on gr-utils/python/blocktool/tests/test_blocktool.py:
```diff
@@ -0,0 +1,187 @@
+                BlockHeaderParser(**test_dict).run
+        # test for invalid header module
+        test_dict['file_path'] = os.path.abspath(
+            os.path.join(self.current_folder, '../../../../gr-utils/python/blocktool/__init__.py'))
```
???

aru31 replied:
This was a debug statement. Sorry for the mess. I have now cleaned up the tests.

skoslowski commented on gr-utils/python/blocktool/tests/test_blocktool.py:
```diff
@@ -0,0 +1,187 @@
+        # test for none type file_path
+        test_dict['file_path'] = None
+        with self.assertRaises(TypeError):
+            BlockHeaderParser(**test_dict).run
```
missing ()?

skoslowski commented on gr-utils/python/gr_blocktool:
```diff
@@ -0,0 +1,32 @@
+""" A tool for parsing header files. """
+
+
+from blocktool.cli.base import cli
```
pull the cli into the blocktool package

shouldn't this be a subpackage of gnuradio

aru31 replied:
It definitely is, this is just a temporary script for testing purposes only.

skoslowski commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,281 @@
+        try:
+            open(file_path, 'r')
+        except TypeError:
+            raise TypeError
```
?

skoslowski reviewed (commented):
high-level feedback (let's consider this part1 of my review)

skoslowski commented on gr-utils/python/blocktool/tests/test_blocktool.py:
```diff
@@ -0,0 +1,187 @@
+        super(TestBlocktoolCore, self).__init__(*args, **kwargs)
+        self.current_folder = os.path.abspath(os.path.dirname(__file__))
+        self.test_dir = os.path.abspath(os.path.join(
+            self.current_folder, '../../../../gr-analog/include/gnuradio/analog'))
```
may define a top-level variable that contains the path to the GR source root

aru31 committed:
Simplify blocktool cli structure

aru31 committed:
Refactor blocktool exception handling

aru31 committed:
Split long blocktool unit-tests

aru31 committed:
Parse message ports from the implementation file

aru31 committed:
Add tests for parsed message port id, update sample json files

aru31 committed:
Add blocktool subdirectory, files in CMakeLists.txt

aru31 committed:
Remove test files to run Blocktool

aru31 committed:
Fix: locates implementation file by traversing the module

aru31 committed:
Integrate blocktool with modtool as an external plugin

aru31 committed:
Create proper formatting of io_signature for yaml files

aru31 committed:
Extend modtool makeyaml command to extend support for blocktool

aru31 committed:
Remove external plugin for modtool support, add blocktool independent script

aru31 committed:
Minor formatiing, change function name due to conflict with modtool function

aru31 committed:
Add support to read and add blocktool special comments in header file

aru31 committed:
Fix: Key Errors, Modify Documentation Reader

primercuervo commented on gr-utils/python/blocktool/cli/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+    __init__.py
+    base.py
+    parseheader.py
+    DESTINATION ${GR_PYTHON_DIR}/gnuradio/blocktool/cli
```
@skoslowski do we want to install the cli files? Or should we just leave them to be run locally from the source code? I assume the latter..

skoslowski replied:
I say, install. At least to have a "python -m gnuradio.blocktool". 

therefore blocktool needs a __main__ module
https://docs.python.org/3/using/cmdline.html#cmdoption-m

I'd start with
```
import sys
from .cli.base import cli

sys.exit(cli())
```
we could then also consider removing gr_blocktool

primercuervo commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,299 @@
+
+
+def dict_representer(dumper, data):
+    """ Representer to represent special OrderedDict """
```
"Representer to represent..." seems a bit redundant

aru31 replied:
Representing with Ordered dict and then using it to convert in YAML parameters is quite easy. The whole blocktool yaml_generator is based on this.

primercuervo replied:
I'm not referring to the implementation or return, but to the grammar of the docstring :)

primercuervo commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,299 @@
+        data['outputs'] = output_signature
+
+    _cpp_templates = [('includes', '#include <gnuradio/{}/{}>'.format(block, self.filename)),
+                      ('declartions', '{}::{}::sptr ${{id}}'.format(block, header)),
```
typo in "declarations"

primercuervo commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,278 @@
+                            self.parsed_data['namespace'].append(
+                                str(_namespace).split('::')[-1].split(' ')[0])
+        except RuntimeError:
+            raise BlockToolException(
```
Why re-raising and not just handling the `RuntimeError` as is? As far as I understand, the `BlockToolException` is not specialized but just a wrapper for the general `exception`

primercuervo reviewed (changes requested):
Just a few comments on the fly that I got to do on a free timeslot. I will add more comments tonight

aru31 committed:
Raise warning in case of conflict in the parsed information and blocktool comments

Commit:
grc: embedded python module show_id flag

It is necessary to have access to the show_id flag for a python module
in grc to be able to access it elsewhere.  This adds the flag when the
module object is created in the block library.

primercuervo commented on gr-utils/python/blocktool/core/outputschema.py:
```diff
@@ -0,0 +1,169 @@
+    "type": "object",
+    "properties": {
+        "namespace": {
+            "description": "List of nested namspaces",
```
typo "namespaces"

primercuervo commented on gr-utils/python/blocktool/core/parseheader.py:
```diff
@@ -0,0 +1,281 @@
+                        self.parsed_data['methods'].append(setter_args.copy())
+            else:
+                self.parsed_data['methods'] = []
+        except RuntimeError:
```
so the result from not having `setters` (or `getters` for lines 238 on) is the same as when there is a `RuntimeError`. Is this on purpose? Any specific reason not to re-raise with your custom exception and/or notify that an exception happened?

aru31 replied:
Yeah. The else statement is not required. I have made the changes.

primercuervo commented on gr-utils/python/blocktool/tests/test_blocktool.py:
```diff
@@ -0,0 +1,199 @@
+        self.assertTrue(valid_input_message_port)
+        self.assertTrue(valid_output_message_port)
+
+    @unittest.skipIf(SKIP_BLOCK_TEST, 'pygccxml not founf, skipping this unittest')
```
Typo "found"

AndersKaloer committed:
modtool: use Boost UTF for 3.8 OOT blocks

modtool was adding CppUnit qa code for gnuradio 3.8. Boost UTF support
was already implemented in #2080, but for some reason not in add.py.
This commit adds the code from #2080, adapted to the new modtool.

Note that it is based on #2592. Fixes #2722, fixes #2697.

AndersKaloer committed:
Add gr::modname namespace to Boost UTF template

hughpyle committed:
Typo

skoslowski commented on gr-utils/python/blocktool/cli/base.py:
```diff
@@ -0,0 +1,126 @@
+)
+def cli():
+    """Block header parsing tool."""
+    pass
```
IMO, plugin support and subcommands are too much. Especially if there is only one...
I propose to strip all the boilerplate stuff and trim this down to one module `cli.py` 

click is fine, but I don't see a need for plugins...

skoslowski commented on gr-utils/python/blocktool/cli/base.py:
```diff
@@ -0,0 +1,126 @@
+        try:
+            if sys.version_info[0] == 2:
+                cmd_name = cmd_name.encode('ascii', 'replace')
+            mod = import_module('gnuradio.blocktool.cli.' + cmd_name)
```
I can't see this working if we run from source, which we always said should be possible. `mod = import_module('.' + cmd_name, package=__package__)` would work, but I say get rid of this (see below)

skoslowski commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,301 @@
+              help='If given, a YAML *file* alongside printed JSON response will be generated')
+@click.option('-c', '--blocktool-comments', is_flag=True,
+              help='blocktool helper comments will be added in the header file')
+@click.option('-f', '--file-confirm', is_flag=True,
```
How about -o/--output for file output instead of stdout

skoslowski commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,301 @@
+@click.command('parseheader',
+               short_help='Generate the parsed output for the header file or directory in a specified format')
+@click.argument('file-path', nargs=1)
+@click.option('-yaml', '--yaml-confirm', is_flag=True,
```
--yaml to dump info as yaml instead of json

aru31 replied:
okay, understood

skoslowski commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,301 @@
+    """
+    kwargs['modtool'] = False
+    if os.path.isfile(kwargs['file_path']):
+        self = BlockHeaderParser(**kwargs)
```
self is not a good name here, how about parser?

skoslowski commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,301 @@
+        run_blocktool(self)
+        click.secho('Header file: {}'.format(self.filename), fg='green')
+        json_generator(self, **kwargs)
+        if kwargs['yaml_confirm']:
```
I shoudl be either json or yaml

skoslowski commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,301 @@
+        raise BlockToolException('Invalid file or directory path.')
+
+
+def json_generator(self, **kwargs):
```
An instance should call itself `self` not a function outside - rename to parser

skoslowski replied:
I also don't see a need for kwargs here

aru31 replied:
Required in line 101

aru31 replied:
> I also don't see a need for kwargs here

Required in line 101

skoslowski commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,301 @@
+        click.secho('Successfully generated {}_{}.json'.format(
+            block, header), fg='green')
+        json_file = os.path.join('.', block+'_'+header + '.json')
+        with open(json_file, 'w') as _file:
```
how about you pass in an open file handle for function to write too, this way it does not need to know the target (stdout or file on disk)

aru31 replied:
I think in case of parsing a directory this would create a problem.

skoslowski replied:
good point. nevermind then

skoslowski commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,301 @@
+            json.dump(self.parsed_data, _file, indent=4)
+    else:
+        print(json.dumps(self.parsed_data, indent=4))
+        click.secho('Successfully parsed header file {}'.format(
```
we wanted the output to be json - so NO extra output...

skoslowski commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,301 @@
+    header = self.filename.split('.')[0]
+    block = self.modname.split('-')[-1]
+    if kwargs['file_confirm']:
+        click.secho('Successfully generated {}_{}.json'.format(
```
usually no output necessary on success for cli only warnings and errors - you can add a --verbose cli option if you want to keep those echos

skoslowski commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,301 @@
+            self.filename), fg='green')
+
+
+def yaml_generator(self):
```
this seems to be the blocktool to grc-yaml code THAT BELONGS IN MODTOOL
the --yaml option in the cli should just dump the info in yaml format instead of json - as was discussed numerous times

skoslowski commented on gr-utils/python/blocktool/cli/parseheader.py:
```diff
@@ -0,0 +1,301 @@
+        data['documentation'] = self.parsed_data['docstring']
+    data['file_format'] = 1
+
+    with open(yml_file, 'w') as yml:
```
I will look at this function after you moved it to modtool.

skoslowski commented on gr-utils/python/blocktool/core/base.py:
```diff
@@ -0,0 +1,50 @@
+    pass
+
+
+class BlockTool(object):
```
check out the abc module in python

skoslowski commented on gr-utils/python/modtool/cli/makeyaml.py:
```diff
@@ -24,14 +24,21 @@
 from .base import common_params, block_name, run, cli_input
 
+from gnuradio.blocktool import BlockHeaderParser
+from gnuradio.blocktool.cli.base import BlockToolException, run_blocktool
```
there must not be any usage of the blocktool cli here - use the API of the blocktool package here not its CLI. pretend it doesn't exist. if that means moving stuff into core then that is what has to happen

skoslowski commented on gr-utils/python/utils/gr_blocktool:
```diff
@@ -0,0 +1,32 @@
+    try:
+        cli()
+    except KeyboardInterrupt:
+        pass
```
I say lets remove this in favor of blocktool.__main__

skoslowski reviewed (changes requested):
This is the first bunch of comments - mostly focusing on the issues surround the core functionality - which is what I'll look at tomorrow in detail.

To be honest I am a bit surprised to find a lot of the issues we had discussed our calls still unsolved. Please address these. A part of me hopes I am looking at an outdated PR here...

aru31 committed:
Remove all the blocktool boilerplate cli code and provide minimal support

aru31 committed:
Remove gr_blocktool script and use blocktool as a python module

aru31 committed:
Major refactoring of the modtool cli structure to support the blocktool API

skoslowski commented on gr-utils/python/blocktool/main.py:
```diff
@@ -0,0 +1,29 @@
+
+def main():
+    """ Call the main function """
+    sys.exit(cli())
```
I think this module can be merged into `__main__`, 
also (minor nitpick) `from .cli import cli should do here =)

skoslowski commented on gr-utils/python/blocktool/cli/cli.py:
```diff
@@ -0,0 +1,138 @@
+                    header), fg='yellow')
+    else:
+        raise BlockToolException(
+            'Invalid directory! No header found to be parsed')
```
since there is only one module left in the cli subpackage, lets make it a single module and move cli/cli.py to cli.py

skoslowski commented on gr-utils/python/blocktool/core/base.py:
```diff
@@ -0,0 +1,52 @@
+        """ Validates the arguments """
+        pass
+
+    def run_blocktool(self):
```
you went for an ABC, cool. But now this should be decorated with `@abc.abstractmethod` instead of raising a NotImplementedError

aru31 committed:
Check for PyGCCXML dependency during build

aru31 committed:
Add README.md for gr-blocktool and remove modtool cli warnings

dl1ksv committed:
Reintroduce default setting in qtgui_chooser

In gnuradio 3.7 it was possible to select an option as default option.
In 3.8 /3.9 the default option is always option 0.

Commit:
qtgui: usrp source gain callbacks not rendered

In the 3.8 gen_uhd_usrp_blocks.py generated an invalid statement for the
set_gain callbacks, so they were never rendered to the output py.

Fixes #2791

argilo committed:
volk: advance volk submodule pointer

This pulls in the fix for the AVX FMA rotator kernel so the rotator
block will work properly on recent Intel CPUs.

mathiasvr committed:
change rounding method to `lround`

grant-cox committed:
Remove circular_file.cc from codebase, cmake, qa, and ConfigChecks #2874

ra1nb0w committed:
codec2: fix support for codec2 version < 0.8

codec2 version < 0.8 doesn't provide codec2/version.h header
therefore cmake discovers if it is available.

fix gr-vocoder build on ubuntu < 19.04 and debian < 10

Closes: https://github.com/gnuradio/gnuradio/pull/2858#issuecomment-549391165

grant-cox committed:
put back checks for io.h in cmake files #2874

ThomasHabets committed:
Use 'const' on more member variables

ThomasHabets committed:
Use C++11 ranged for and ranged insert instead of manual loops

Commit:
Document the statistical distribution that is used to simulate impulse noise.

karel committed:
grc: remove deprecated encoding argument from json.loads

mormj committed:
Merge pull request #2778 from dl1ksv/patch-18

Reintroduce default setting in qtgui_chooser

mormj committed:
Merge pull request #2729 from hughpyle/master

grc: fix type in missing xterm warning

duggabe committed:
docs: revise install instructions (#2923)

* revise install instructions to point to gnuradio/pybombs and wiki installGR

* revise quick start instuctions

* add details to quick start

* add subtitle; minor rewording
